python/cudf/cudf/_lib/pylibcudf/expressions.pxd

```diff
@@ -1,36 +1,29 @@
-# Copyright (c) 2022-2024, NVIDIA CORPORATION.
-
-from libc.stdint cimport int32_t, int64_t
+# Copyright (c) 2024, NVIDIA CORPORATION.
 from libcpp.memory cimport unique_ptr
+from libcpp.string cimport string
 
 from cudf._lib.pylibcudf.libcudf.expressions cimport (
-    column_reference,
+    ast_operator,
     expression,
-    literal,
-    operation,
-)
-from cudf._lib.pylibcudf.libcudf.scalar.scalar cimport (
-    numeric_scalar,
-    scalar,
-    string_scalar,
-    timestamp_scalar,
+    table_reference,
 )
+from cudf._lib.pylibcudf.libcudf.scalar.scalar cimport scalar
 
 
 cdef class Expression:
     cdef unique_ptr[expression] c_obj
 
-
 cdef class Literal(Expression):
     cdef unique_ptr[scalar] c_scalar
 
-
 cdef class ColumnReference(Expression):
     pass
 
-
 cdef class Operation(Expression):
-    pass
+    # Hold on to the input expressions so
+    # they don't get gc'ed
+    cdef Expression right
```

Comment:
There was a use-after-free here previously, but it didn't get caught I think because the inputs to Operations weren't gc'ed yet.

Reply:
In general, if the C++ side takes a reference, we need to hold on to the object on the Python side, right?

Reply:
Yup we do, that's a good catch.